### PR TITLE
[RFC] Fix #5984

### PR DESCRIPTION
**Pull request by sander2**
Rebase & squash of the previous PR. Can I send this to Neovim to ask for help on the testing?

Doesn't have to be perfect, per se:
 > @ sander2: I'll need to cleanup things a bit before making an actual PR. I'll make the PR tomorrow, and ask for help there.
@ justinmk: Sure no problem. But, messy PRs are OK
(if not then i've been doing it wrong)
just mark it [WIP]

### CODE REVIEW

File: src/nvim/msgpack_rpc/channel.c [hunk 1]
```diff
@@ -409,38 +408,40 @@ static void handle_request(Channel *channel, msgpack_object *request)
                channel->id);
       call_set_error(channel, buf);
     }
-    return;
   }
+  else {
```
Comment (Hjdskes) [inline src/nvim/msgpack_rpc/channel.c]: Should be `} else {`

File: src/nvim/api/private/helpers.c [hunk 2]
```diff
@@ -971,3 +970,19 @@ static void set_option_value_err(char *key,
...
+
+void _api_set_error(Error *err, ErrorType errType, const char *format, ...)
+{
+    va_list args1;
+    va_start(args1, format);
+    va_list args2;
+    va_copy(args2, args1);
+    int len = vsnprintf(NULL, 0, format, args1);
+    va_end(args1);
+	err->msg = xmalloc(len+1);
```
Comment (Hjdskes) [inline src/nvim/api/private/helpers.c]: Indentation, below too.

File: src/nvim/api/private/helpers.c [hunk 3]
```diff
@@ -971,3 +970,19 @@ static void set_option_value_err(char *key,
     api_set_error(err, Exception, "%s", errmsg);
   }
 }
+
+void _api_set_error(Error *err, ErrorType errType, const char *format, ...)
+{
+    va_list args1;
+    va_start(args1, format);
+    va_list args2;
```
Comment (Hjdskes) [inline src/nvim/api/private/helpers.c:978]: Why two `va_list`s? Can't you reuse a single one?

File: src/nvim/api/private/helpers.h [hunk 4]
```diff
@@ -8,14 +8,9 @@
 #include "nvim/memory.h"
 #include "nvim/lib/kvec.h"
 
+void _api_set_error(Error *err, ErrorType errType, const char *format, ...);
```
Comment (Hjdskes) [inline src/nvim/api/private/helpers.h:11]: What's the idea behind this `_api_set_error` function name? I get that in some codebases the underscore implies it's private, but does Neovim actually do this too?

File: src/nvim/api/private/helpers.c [hunk 5]
```diff
@@ -971,3 +970,19 @@ static void set_option_value_err(char *key,
...
+    va_copy(args2, args1);
+    int len = vsnprintf(NULL, 0, format, args1);
+    va_end(args1);
+	err->msg = xmalloc(len+1);
+    vsprintf(err->msg, format, args2);
+    va_end(args2);
+	
+	err->set = true;
```
Comment (Hjdskes) [inline src/nvim/api/private/helpers.c]: Are you sure you can unconditionally set this to `true`? I see only one such instance where you actually removed `err->set = true` (channel.c, line 222); what about the others?

File: src/nvim/msgpack_rpc/channel.c [hunk 6]
```diff
@@ -217,9 +217,8 @@ Object channel_send_call(uint64_t id,
           && (array.items[0].data.integer == kErrorTypeException
               || array.items[0].data.integer == kErrorTypeValidation)
           && array.items[1].type == kObjectTypeString) {
-        err->type = (ErrorType) array.items[0].data.integer;
-        xstrlcpy(err->msg, array.items[1].data.string.data, sizeof(err->msg));
-        err->set = true;
+		_api_set_error(err, (ErrorType) array.items[0].data.integer, "%s",
```
Comment (Hjdskes) [inline src/nvim/msgpack_rpc/channel.c]: Indentation

File: src/nvim/terminal.c [hunk 7]
```diff
@@ -1217,13 +1218,14 @@ static bool is_focused(Terminal *term)
...
   do { \
-    Error err; \
+    Error err = ERROR_INIT; \
     /* Only called from terminal_open where curbuf->terminal is the */ \
     /* context  */ \
     o = dict_get_value(curbuf->b_vars, cstr_as_string(k), &err); \
     if (o.type == kObjectTypeNil) { \
       o = dict_get_value(&globvardict, cstr_as_string(k), &err); \
     } \
+	xfree(err.msg); \
```
Comment (Hjdskes) [inline src/nvim/terminal.c]: Indentation

**Review by Hjdskes — CHANGES_REQUESTED** [review]
Some style fixes required and some questions from my side.

One more question: can you be reasonably sure you have converted all error cases?

File: src/nvim/api/private/helpers.c (hunk 3 above)
Comment (sander2) [reply src/nvim/api/private/helpers.c:978]: No. I got my example from [here](http://en.cppreference.com/w/c/io/vfprintf). A `va_list` is a struct which contains a a pointer to the 'current' parameter. When an argument is read, this pointer increases. The `va_end` could potentially restore the `va_args` to the initial state, but this is [not](http://en.cppreference.com/w/cpp/utility/variadic/va_end) guaranteed by the standard.

**Comment by sander2** [comment]
I'll fix the stylistic errors. And yes, I'm reasonably sure I got all references; this was explained in #2. You're right about the function name, I should rename it.

File: src/nvim/api/private/helpers.c (hunk 5 above)
Comment (sander2) [reply src/nvim/api/private/helpers.c]: The function replaces the previous `api_set_error` define, where it was also set to true. All new occurrences of `_api_set_error` I added replaced code that has manual `err->set = true;` lines, if I'm not mistaken.

**Comment by Hjdskes** [comment]
I think we can open a WIP PR to Neovim now to gather their feedback, so I'm merging this PR.

**Pull request merged**